test/hotspot/gtest/runtime/test_atomic.cpp

```diff
@@ -199,3 +199,97 @@ TEST(AtomicEnumTest, scoped_enum) {
   Support().test_cmpxchg(B, C);
   Support().test_xchg(B, C);
 }
+
+template<typename T>
+struct AtomicBitopsTestSupport {
+  volatile T _test_value;
+
+  // At least one byte differs between _old_value and _old_value op _change_value.
+  static const T _old_value =    static_cast<T>(UCONST64(0x7f5300007f530000));
+  static const T _change_value = static_cast<T>(UCONST64(0x3800530038005300));
+
+  AtomicBitopsTestSupport() : _test_value(0) {}
+
+  void fetch_then_and() {
+    Atomic::store(&_test_value, _old_value);
+    T expected = _old_value & _change_value;
+    T result = Atomic::fetch_then_and(&_test_value, _change_value);
+    EXPECT_EQ(_old_value, result);
+    EXPECT_EQ(expected, Atomic::load(&_test_value));
+  }
+
+  void fetch_then_or() {
+    Atomic::store(&_test_value, _old_value);
+    T expected = _old_value | _change_value;
+    T result = Atomic::fetch_then_or(&_test_value, _change_value);
+    EXPECT_EQ(_old_value, result);
+    EXPECT_EQ(expected, Atomic::load(&_test_value));
+  }
+
+  void fetch_then_xor() {
+    Atomic::store(&_test_value, _old_value);
+    T expected = _old_value ^ _change_value;
+    T result = Atomic::fetch_then_xor(&_test_value, _change_value);
+    EXPECT_EQ(_old_value, result);
+    EXPECT_EQ(expected, Atomic::load(&_test_value));
+  }
+
+  void and_then_fetch() {
+    Atomic::store(&_test_value, _old_value);
+    T expected = _old_value & _change_value;
+    T result = Atomic::and_then_fetch(&_test_value, _change_value);
+    EXPECT_EQ(expected, result);
+    EXPECT_EQ(expected, Atomic::load(&_test_value));
+  }
+
+  void or_then_fetch() {
+    Atomic::store(&_test_value, _old_value);
+    T expected = _old_value | _change_value;
+    T result = Atomic::or_then_fetch(&_test_value, _change_value);
+    EXPECT_EQ(expected, result);
+    EXPECT_EQ(expected, Atomic::load(&_test_value));
+  }
+
+  void xor_then_fetch() {
+    Atomic::store(&_test_value, _old_value);
+    T expected = _old_value ^ _change_value;
+    T result = Atomic::xor_then_fetch(&_test_value, _change_value);
+    EXPECT_EQ(expected, result);
+    EXPECT_EQ(expected, Atomic::load(&_test_value));
+  }
+
+#define TEST_BITOP(name) { SCOPED_TRACE(XSTR(name)); name(); }
+
+  void operator()() {
+    TEST_BITOP(fetch_then_and)
+    TEST_BITOP(fetch_then_or)
+    TEST_BITOP(fetch_then_xor)
+    TEST_BITOP(and_then_fetch)
+    TEST_BITOP(or_then_fetch)
+    TEST_BITOP(xor_then_fetch)
+  }
+
+#undef TEST_BITOP
+};
+
+template<typename T>
+const T AtomicBitopsTestSupport<T>::_old_value;
+
+template<typename T>
+const T AtomicBitopsTestSupport<T>::_change_value;
+
+TEST(AtomicBitopsTest, int32) {
+  AtomicBitopsTestSupport<int32_t>()();
+}
+
+TEST(AtomicBitopsTest, int64) {
+  AtomicBitopsTestSupport<int64_t>()();
+}
```

Comment:
These 64-bit tests would fail on 32-bit systems, are they not? They are outside the requirements for the contract to only support `sizeof(int)` and `sizeof(void*`). Probably just `#ifdef _LP64` these?

Reply:
Well-spotted.  I forgot to conditionalize for `_LP64` as is done for other tests in this file.